handlers.go

```diff
@@ -58,21 +58,32 @@ func BeginAuth(w http.ResponseWriter, r *http.Request) {
     }
 }
 
-func getSingleSecQues(roll string) string {
+func getSingleSecQues(roll string,c chan string) {
+    defer func() {
+        if (recover() != nil) {
+            log.Printf("Exiting gracefully\n")   // in case panic occurs while writing to a closed channel
+        }
+    }()
     v := url.Values{}
     v.Set("user_id", roll)
     resp, _ := http.PostForm(ERP_SECRET_QUES_URL, v)
     body, _ := ioutil.ReadAll(resp.Body)
-    return string(body)
+    c <- string(body)
+    
 }
 
+
 func getSecurityQuestions(roll string) []string {
     allSecQues := []string{}
+    data := make(chan string)
 
-    // Perform upto 30 tries to get the 3 unique secret questions from ERP
-    for i := 1; i < 30; i++ {
-        secQues := getSingleSecQues(roll)
-        log.Printf("Run %d, Got %s", i, secQues)
+    for i := 1 ;i <= 30; i++ {     // Perform upto 30 tries to get the 3 unique secret questions from ERP
+        go getSingleSecQues(roll,data)
```

Comment:
Actually, instead of passing the channel, it's perhaps better to do this with an anon function:

```go
go func() { data <- getSingleSecQues(roll) }
```

This let's us use the `getSingleSecQues` function even without channels (in a synchronous situation). I would actually like to find out what the "Go way" of doing this is. Perhaps, @kshitij10496 can help us with this? (Sorry for the tag out of the blue! 🙂 )

Reply:
Ah, okay, thanks for your input Kshitij. I had a look at the Generator Pattern, and it seems like we would have code that looks like the following:

```go
func getSecurityQues(roll string) string {
    ...
    return "<sec_ques>"
}

func wrapper1(roll string) <-chan {
    c := make (chan string)
    for i := 0; i < 30; i++ {
        go func() { c<-getSecurityQues(roll) }
    }
    return c;
}

// in main

c := wrapper1(roll_num)
for (i = 0; i < 30; i++) {
    ques := <-c
    ...
}
```

I don't really see how this code is easier to read than before. I would much rather have the two for-loops in the main function, thus reducing one level of function calls and make it more readable (`wrapper1` serves no purpose except for declaring the channel and returning it)

```go
// inside main
c := make(chan string)
for i := 0; i < 30; i++ {
    go func() { c <- getSecurityQues(roll) }
}

for i = 0; i < 30; i++ {
    ques := <-c
    ...
}
```

What do you think?

_For the future reader, who might be wondering why we are nit picking over such details_ 

P.S I am new to this language and there seem to be several patterns that can be used to get to the same implementation goal. More discussion at this early stage of language learning is very beneficial! 🙂

Reply:
@icyflame I too feel the same about the wrapper function. And the two `for` loops  seems more readable to me.  But a discussion on the pros and cons of both ways is really worth having.